compiler/packages/babel-plugin-react-compiler/src/Validation/ValidateNoSetStateInPassiveEffects.ts

```diff
@@ -0,0 +1,152 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import {CompilerError, ErrorSeverity} from '../CompilerError';
+import {
+  HIRFunction,
+  IdentifierId,
+  isSetStateType,
+  isUseEffectHookType,
+  Place,
+} from '../HIR';
+import {eachInstructionValueOperand} from '../HIR/visitors';
+
+/**
+ * Validates against calling setState in the body of a *passive* effect (useEffect),
+ * while allowing calling setState in callbacks scheduled by the effect.
+ *
+ * Calling setState during execution of a useEffect triggers a re-render, which is
+ * often bad for performance and frequently has more efficient and straightforward
+ * alternatives. See https://react.dev/learn/you-might-not-need-an-effect for examples.
+ */
+export function validateNoSetStateInPassiveEffects(fn: HIRFunction): void {
+  const setStateFunctions: Map<IdentifierId, Place> = new Map();
```

Comment:
This is profoundly nitpicky given how hard it is to construct an example of this, but we might want to first detect all the setStateFunctions in a component before taking another pass and detecting errors, because of setState functions that might be [hoisted](https://react-compiler-playground-git-gh-josephsavo-f70b91-fbopensource.vercel.app/#N4Igzg9grgTgxgUxALhAejQAgAIDMA2EA7jgG4CG+AlgCbkAuCAchAMoL2v0MICSAdgAVyYMFVIIAorlwI49MAB1+y3FH7yqEfpgBiECAAoAlJmDLMmfB0wAPTAF5MJxwD4zmMLZcBfANwWmHDaYPSYANqhPAA0nhxcPAC6jphQYAgJjIYADMYBOkEhYV4p6ZzcjPmWaVIycvSGLg7uwHaGAIymPnmBMBywOlGVyj7KID5AA)

Reply:
Per offline discussion the same issue applies to several other validations, let's land this and do a follow-up pass to apply the same two-phase approach to all the validations (and maybe take that as a chance to create more of a framework for validations)